spec/p2p/p2p-top-level.md

```diff
@@ -0,0 +1,310 @@
+*** This is the beginning of an unfinished draft. Don't continue reading! ***
+
+# << Tendermint P2P >>
+
+<!--- > Rough outline of what the component is doing and why. 2-3 paragraphs 
+--->
+
+The Tendermint consists of multiple protocols, namely,
+- Consensus
+- Mempool
+- Evidence
+- Blocksync
+- Statesync
+
+that each plays a role in making sure that validators can produce blocks. These protocols are implemented in so-called reactors (one for each protocol) that encode two functionalities:
+
+- Protocol logic (controlling the local state of the protocols and deciding what messages to send to others, e.g., the rules we find in the arXiv paper)
+
+- Communication. Message exchange with other nodes (Gossip)
+
+Tendermint (as many classic BFT algorithms) have an all-to-all communication pattern (e.g., every validator sends a `precommit` to every other validator). Naive implementations, e.g., maintaining a channel between each of the *N* validators is not scaling to the system sizes of typical Cosmos blockchains (e.g., N = 200 validator nodes + seed nodes + sentry nodes + other full nodes). There is the fundamental necessity to restrict the communication.
+
+The design decision is to use an overlay network. Instead of having *N* connections, each node only maintains a relatively small number (bounded by constants, say 10 to 50). In principle, this allows to implement more efficient communication (e.g., gossiping), provided that with this small number of connections per node, the system as a whole stays connected. This overlay network 
```

Comment:
We might need to define "gossip" more broadly and possibly precisely here. Gossip should not necessarily need a network overlay, as the original proposal assumed interactions with nodes chosen at random at each round of the communication protocol. Another thing that probably I should clarify is that by "gossip" we somehow also mean synchronization of states. This is what part of the protocols (state and block sync in particular) do, while they don't use what is known as gossip communication: they are "pure" peer-to-peer protocols.

Reply:
Thanks for the comment. I didn't realize this, but it seems that Consensus gossip is a property stated in the arxiv paper, but the solution is in fact not a gossip-based solution but a peer-to-peer protocol. We should also incorporate this in the write-up.

Reply:
Let's discuss this point synchronously.